### ISSUE TITLE
iadmin rebalance error handing 

### ISSUE THREAD

**Issue by kript** [irods/irods#3585]
Hi folks,

We’ve seen an uptick in the number of times we’ve had to restart the rebalance recently, so I thought I’d check that the following situations are handled by the code for 4.1.11, possibly by reporting but not crashing unless a better option (check and retry perhaps, is available);

1.	File removed during rebalance
2.	Individual replication failed due to;
a.	Network issues (`SYS_COPY_LEN_ERR`, `SYS_HEADER_READ_LEN_ERR`)
b.	Resource offline at the time of rebalance trying to access file (`UNIX_FILE_OPEN_ERR`, `UNIX_FILE_WRITE_ERR`)
c.	Checksum issue (`USER_CHKSUM_MISMATCH`) – not sure what’s causing this ATM, more information in the logs might help
d.	iRES offline (`USER_SOCK_CONNECT_ERR`)

A common thread here also appears to be the rebalance not checking the status of a resource before trying to rebalance to/from it – we are somewhat aggressively updating our iRES server estate firmware at the moment, and this means marking all the resources on one as down when there is no traffic to it (as evidenced by the the output of `ips -H ${HOSTNAME}` on the server).

If we’re doing it wrong related to bringing the iRES down, please do say. We’re currently making use of `admin modresc XXX status down`, should we be combining this with `irods-grid pause <ires_server>`?

Cheers

John

**Comment by trel** [comment]
Regarding...

1) files removed/gone... You can probably reduce the window of opportunity for that occurring by lowering the number of files in the loop

```
pep_resource_rebalance_pre(*OUT) {
    *OUT="replication_rebalance_limit=200";
}
```

  https://github.com/irods/irods_docs/commit/62b7d1dee685065af4bf1a051ac7a5bafe767ae0#diff-fc84350ac42f2daf48445f4fb79b30f3R133

2) Each replication event will do a hierarchy resolution as part of the preparation for the replication itself.  So it should already be 'checking' the status of the resource involved when it checks that hierarchy.  Logging lots of information is https://github.com/irods/irods/issues/3476 and is slated for 4.1.11.

Aggressively changing the state of the system components while rebalance is running was not an original use case, but I agree, better logging and continuation should allow things to make progress in spite of all the hurdles.

Marking the resource down in the catalog is fine.  Pausing the system on the server itself will stop all new traffic to that server, not just affecting a singular resource hanging off that server.  You could use `irods-grid` if you want, but that may introduce more moving parts to an already complicated set of scenarios.  I'd say just use `modresc` as you are for now (unless you're looking to reduce the number of commands to run more than reducing the number of variables :) )

**Comment by trel** [comment]
@kellerb @kript is there anything actionable here?   can we close this?

**Comment by kript** [comment]
I think other tickets have addressed a lot of this, but I'm not sure this one has got anything against it;

> Aggressively changing the state of the system components while rebalance is running was not an original use case, but I agree, better logging and continuation should allow things to make progress in spite of all the hurdles.

So my worry is that its looking like we need to run a rebalance perhaps daily. The system state _is more likely to_ change the larger amount of data ingested while a long running rebalance is ongoing. Even if running once a day reduced the window if time the rebalance runs for, I would argue it should check the system state - obviously theres nothing that can be done about a server going AWOL without being marked down and those errors should be generated, just not for resources where "irods" knows the resource is down, if that makes sense...
Perhaps if the check for resource down was done every invocation of `pep_resource_rebalance_pre` or something?

**Comment by trel** [comment]
As I understand it - we're already checking the resource states when selecting a replication target, for every replication attempt (for every data object).  This is much more often than for each invocation of rebalance at the system level.

If rebalance needs to be run every day, and there is lots of work to do to get things 'back' into balance every day... we should probably spend time looking at the tree/policy/dataflow itself to determine why things are so dynamic/unstable.

**Comment by keithj** [comment]
Hi @kript, @trel,

My understanding of what's happening is that since our pipelines are uploading 24x7, when we encounter hardware errors or network glitches on the LAN or WAN affecting a resource server, we are likely to see an effect on some upload operations. If the resource servers for a file's replicates are affected differentially, then we are likely to see unbalanced replicates.

To put some numbers to that, in the last 3 months we uploaded >15 miliion files (judged by DATA_ID count), which I make to be about 115 per minute, on average. We had to rebalance ~30k files from that period i.e. a 0.2% failure rate. That seems high, but these are not idependent events - once a server has an issue it affects a lot of files at once, even if the cause lasts for only a few seconds.

Keith
